File: leetcode/algorithms/680_valid_palindrome_2/main.py
class ValidPalindromeII:
    def validPalindrome(self, s: str) -> bool:
        left: int = 0
        right: int = len(s) - 1

        while left < right and s[left] == s[right]:
            left += 1
            right -= 1

        left1: int = left + 1
        right1: int = right

        left2: int = left
        right2: int = right - 1

        while left1 < right1 and s[left1] == s[right1]:
            left1 += 1
            right1 -= 1

        while left2 < right2 and s[left2] == s[right2]:
            left2 += 1
            right2 -= 1

        return left1 >= right1 or left2 >= right2


    # Solution
    def solution(self, s):
        i, j = 0, len(s) - 1

        while i < j:
            if s[i] == s[j]:
                i += 1
                j -= 1
            else:
                return self.validPalindromeUtil(s, i + 1, j) or self.validPalindromeUtil(s, i, j - 1)
        return True

    def validPalindromeUtil(self, s, i, j):
        while i < j:
            if s[i] == s[j]:
                i += 1
                j -= 1
            else:
                return False

        return True

File: leetcode/algorithms/680_valid_palindrome_2/main.cpp
#include <string>
using namespace std;

class ValidPalindromeII {
public:
    bool validPalindrome(string s) {
        int left = 0, right = s.size() - 1, deleted = 0;

        while (left < right) {
            if (s.at(left) == s.at(right)) {
                left++;
                right--;
            } else {
                if (!deleted) {
                    if (s.at(left + 1) == s.at(right)) {
                        left++;
                        deleted = 1;
                    } else {
                        break;
                    }
                } else {
                    break;
                }
            }
        }

        if (left >= right) {
            return true;
        } else {
            left = 0, right = s.size() - 1, deleted = 0;
        }

        while (left < right) {
            if (s.at(left) == s.at(right)) {
                left++;
                right--;
            } else {
                if (!deleted) {
                    if (s.at(left) == s.at(right - 1)) {
                        right--;
                        deleted = 1;
                    } else {
                        break;
                    }
                } else {
                    break;
                }
            }
        }

        if (left >= right) {
            return true;
        }

        return false;
    }


    // Solution
    bool solution(string s) {
        for (int left = 0, right = s.size() - 1; left < right; left++, right--) {
            if (s.at(left) != s.at(right)) {
                int left1 = left + 1, right1 = right;
                int left2 = left, right2 = right - 1;
                while (left1 < right1 && s.at(left1) == s.at(right1)) {
                    left1++;
                    right1--;
                }
                while (left2 < right2 && s.at(left2) == s.at(right2)) {
                    left2++;
                    right2--;
                }
                return left1 >= right1 || left2 >= right2;
            }
        }
        return true;
    }
};
